### ISSUE TITLE
from_builtins, but for dict-like objects (asyncpg.Record)

### ISSUE THREAD

**Issue by Gjgarr** [jcrist/msgspec#423]
### Description

# Description

In a similar fashion to #416, I would like `msgspec.from_builtins` to be able to handle dict-like objects (specifically in this case, `asyncpg.Record`s)

# Example Code

Requires a postgres database running on localhost to replicate due to being unable to construct `asyncpg.Record`s easily

```py
import asyncpg
import msgspec

conn = await asyncpg.connect(...)  # creds here
```
```py
# `db_record` will be a asyncpg.Record with nested Record's
async with conn.transaction():
    await conn.execute("""
    CREATE TEMP TABLE inner_t AS
      SELECT * FROM (
        VALUES (1, 'x'), (2, 'y')
      ) AS _ (c, d)
    """)
    db_record = await conn.fetchrow("""
    SELECT outer_t.a
         , outer_t.b
         , ARRAY_AGG(inner_t.* ORDER BY d DESC) AS inners
      FROM (VALUES (1, 'b')) AS outer_t (a, b), inner_t
    GROUP BY 1,2
    """)
```

```py
print(db_record)
# <Record a=1 b='b' inners=[<Record c=2 d='y'>, <Record c=1 d='x'>]>
```

```py
class InnerEntity(msgspec.Struct):
    c: int
    d: str
class OuterEntity(msgspec.Struct):
    a: int
    b: str
    inners: list[InnerEntity]

conv_obj = msgspec.from_builtins(db_record, type=OuterEntity)
# ValidationError
```

# Other Libraries

This feature can be accomplished in cattrs via `cattrs.structure(db_record, OuterEntity)`.

I would like a similar means of structuring these kinds of objects in msgspec so that I can fully migrate.

# Other Considerations

`dict(db_record)` - while this will convert the top-level object to something `from_builtins` can mostly handle, it will not convert the nested objects; I would like msgspec to handle the entire conversion process.

`asyncpg.Record` does not directly inherit from `dict` but `hasattr(db_record, "__getitem__")` is `True`.
Perhaps `from_builtins` could check for the dunder before throwing the validationerror and treat the object as a dict in such case?

Another consideration is `asyncpg.Record` implements `__getitem__`, `get()`, `items()`, `keys()`, `values()` -- but other objects that implement `__getitem__` may not support the other methods that one would expect.

# Ideal usage

```py
msgspec.from_builtins(db_record, type=OuterEntity)
```

Ideally `from_builtins` should recognize the object (and its nested counterparts) as capable of being treated like a dict.

If another kwarg is needed like `attributes`, that works.

**Comment by jcrist** [comment]
Thanks for opening this issue, this is definitely in scope, and should be handleable without requiring a new keyword argument to `from_builtins`. Tossing it on the todo list.

**Comment by jcrist** [comment]
This has been resolved by #427. I suspect there may be a couple weeks before this is part of a release (the spur of feature requests for `from_builtins` has made me want to fix them all for the same release). Until an official release, you can try this out by [installing from GitHub](https://jcristharif.com/msgspec/install.html#installing-from-github).
